# Project 2
### CS 5/7394 - Applied Machine Learning

- **Due** - March 11 @ 11:59 pm pushed to Github repo
- **Teams** - You can do this project solo or in pairs.  Not 3, not 4 not 5... Max of 2. If a 5394 student pairs with a 7394 student, the pair needs to do the 7394 work. 

Below are 6 Kaggle Datasets.  You will choose 1 to work with for this project. 

- [Airfare Prediction Dataset](https://www.kaggle.com/zwartfreak/airline-fare-prediction)
- [Chinese Rest Holiday Dataset](https://www.kaggle.com/holoong9291/chinese-rest-holiday-dataset-2020-to-2022)
- [Jigsaw Toxic Comment Classification Challenge](https://www.kaggle.com/julian3833/jigsaw-toxic-comment-classification-challenge?select=train.csv)
- [Latest Covid 19 Dataset Worldwide](https://www.kaggle.com/sandhyakrishnan02/latest-covid-19-dataset-worldwide)
- [Trains](https://www.kaggle.com/vahidehdashti/machinelearningdatabase-trains?select=trains-original.data)
- [Football Data top 5 Leagues](https://www.kaggle.com/sanjeetsinghnaik/football-data-top-5-leagues)

Merging disparate datasets is a staple of the data exploration process.  Therefore, for which ever data set above that you choose, you will need to independently find **an additional** dataset to merge with your selection.  The only requirement is that it add to the richness of the original dataset. Students in the 7000-level version of the class need to find two additional data sets to merge with the original selection. 

_Note_: If you want to start with a different data set, you need to get Fontenot's OK first.  

### Your Tasks

Below, there are cells that provide directions on what to do for the project.  

You can insert as many cells between the ones below as you'd like, but please **Do NOT** change the cells already provided. 

### Part 1 - Getting Started

- Import libraries
- Load original Data (which ever one you chose from the provided list) into a data frame. 
- Load your additional data set(s) into a data frame. 
- In a markdown cell, provide a brief description of your the data sets you've chosen to work with.  
- Develop a list of 3 - 4 questions that you hope to be able to answer after the exploration of the data and write them in this section.  (Relationship Questions)

In [1]:
import pandas

In [4]:
data_orig = pandas.read_csv("data/data-orig.csv")

<strong>Original Data Description</strong>

<strong>Name</strong>: Huge Dataset - COVID 19 worldwide case

<strong>Description:</strong> This dataset contains COVID-19 details by country until January 8th, 2022.

<strong>Columns</strong>:
- iso_code (categorical): The ISO code for a country
- continent (categorical): The continent the country is on
- location (categorical): The name of the country
- date (categorical): Date the data was collected
- total_cases (numerical): Total confirmed cases of COVID-19 in the country
- new_cases (numerical): Total confirmed cases of CAVID-19 since the last data collection
- new_cases_smoothed (numerical): new_cases smoothed to 7 days
- total_deaths (numerical): Total deaths due to COVID-19
- new_deaths (numerical): Total new deaths due to COVID-19 since the last data collection
- new_deaths_smoothed (numerical): new_deaths column smoothed to 7 days
- total_cases_per_million (numerical): Total confirmed COVID-19 cases per million people in the population
- new_cases_per_million (numerical): Total new confirmed COVID-19 cases per million people in the population
- new_cases_smoothed_per_million (numerical): new_cases_per_million column smoothed to 7 days
- total_deaths_per_million (numerical): Total number of deaths per million people in the population
- new_deaths_per_million (numerical): Total number of new deaths per million people in the population
- new_deaths_smoothed_per_million (numerical): new_deaths_per_million smoothed to 7 days
- reproduction_rate (numerical): Real-time estimate of the effective reproduction rate (R) of COVID-19
- icu_patients (numerical): Number of ICU patients on a given day
- icu_patients_per_million (numerical): Number of ICU patients per mullion people in the population on a given day
- hosp_patients (numerical): Number of hospitalized patients on a given day
- hosp_patients_per_million (numerical): Number of hospitalized patients per million people in the population on a given day
- weekly_icu_admissions (numerical): Number of COVID-19 patients newly admitted to intensive care units (ICUs) in a given week
- weekly_icu_admissions_per_million (numerical): Number of COVID-19 patients newly admitted to intensive care units (ICUs) in a given week per million people
- weekly_hosp_admissions (numerical): Number of COVID-19 patients newly admitted to hospitals in a given week
- weekly_hosp_admissions_per_million (numerical): Number of COVID-19 patients newly admitted to hospitals in a given week per million people
- new_tests (numerical): New tests for COVID-19 (only calculated for consecutive days)
- total_tests (numerical): Total tests for COVID-19
- total_tests_per_thousand (numerical): Total tests for COVID-19 per 1,000 people
- new_tests_per_thousand (numerical): New tests for COVID-19 per 1,000 people
- new_tests_smoothed (numerical): New tests for COVID-19 (7-day smoothed)
- new_tests_smoothed_per_thousand (numerical): New tests for COVID-19 (7-day smoothed) per 1,000 people
- positive_rate (numerical): The share of COVID-19 tests that are positive, given as a rolling 7-day average (this is the inverse of testspercase)
- tests_per_case (numerical): Tests conducted per new confirmed case of COVID-19, given as a rolling 7-day average (this is the inverse of positiverate)
- tests_units (categorical): Units used by the location to report its testing data
- total_vaccinations (numerical): Total number of COVID-19 vaccination doses administered
- people_vaccinated (numerical): Total number of people who received at least one vaccine dose
- people_fully_vaccinated (numerical): Total number of people who received all doses prescribed by the vaccination protocol
- total_boosters (numerical): Total number of COVID-19 vaccination booster doses administered
- new_vaccinations (numerical): New COVID-19 vaccination doses administered (only calculated for consecutive days)
- new_vaccinations_smoothed (numerical): New COVID-19 vaccination doses administered (7-day smoothed)
- total_vaccinations_per_hundred (numerical): Total number of COVID-19 vaccination doses administered per 100 people in the total population
- people_vaccinated_per_hundred (numerical): Total number of people who received at least one vaccine dose per 100 people in the total population
- people_fully_vaccinated_per_hundred (numerical): Total number of people who received all doses prescribed by the vaccination protocol per 100 people in the total population
- total_boosters_per_hundred (numerical): Total number of COVID-19 vaccination booster doses administered per 100 people in the total population
- new_vaccinations_smoothed_per_million (numerical): New COVID-19 vaccination doses administered (7-day smoothed) per 1,000,000 people in the total population
- new_people_vaccinated_smoothed (numerical): Daily number of people receiving their first vaccine dose (7-day smoothed)
- new_people_vaccinated_smoothed_per_hundred (numerical): Daily number of people receiving their first vaccine dose (7-day smoothed) per 100 people in the total population
- stringency_index (numerical): Government Response Stringency Index: composite measure based on 9 response indicators including school closures, workplace closures, and travel bans, rescaled to a value from 0 to 100 (100 = strictest response)
- population (numerical): Population of the country
- population_density (numerical): Number of people divided by land area, measured in square kilometers, most recent year available
- median_age (numerical): Median age of the population, UN projection for 2020
- aged_65_older (numerical): Share of the population that is 65 years and older, most recent year available
- aged_70_older (numerical): Share of the population that is 70 years and older in 2015
- gdp_per_capita (numerical): Gross domestic product at purchasing power parity (constant 2011 international dollars), most recent year available
- extreme_poverty (numerical): Share of the population living in extreme poverty, most recent year available since 2010
- cardiovasc_death_rate (numerical): Death rate from cardiovascular disease in 2017 (annual number of deaths per 100,000 people)
- diabetes_prevalence (numerical): Diabetes prevalence (% of population aged 20 to 79) in 2017
- female_smokers (numerical): Share of women who smoke, most recent year available
- male_smokers (numerical): Share of men who smoke, most recent year available
- handwashing_facilities (numerical): Share of the population with basic handwashing facilities on-premises, most recent year available
- hospital_beds_per_thousand (numerical): Hospital beds per 1,000 people, most recent year available since 2010
- life_expectancy (numerical): Life expectancy at birth in 2019
- human_development_index (numerical): A composite index measuring average achievement in three basic dimensions of human development—a long and healthy life, knowledge, and a decent standard of living.
- excess_mortality_cumulative_absolute (numerical): Cumulative difference between the reported number of deaths since 1 January 2020 and the projected number of deaths for the same period based on previous years.
- excess_mortality_cumulative (numerical): Percentage difference between the cumulative number of deaths since 1 January 2020 and the cumulative projected deaths for the same period based on previous years.
- excess_mortality (numerical): Percentage difference between the reported number of weekly or monthly deaths in 2020–2021 and the projected number of deaths for the same period based on previous years.
- excess_mortality_cumulative_per_million (numerical): Cumulative difference between the reported number of deaths since 1 January 2020 and the projected number of deaths for the same period based on previous years, per million people.


<strong>In total</strong>:
- 5 categorical
- 65 numerical

<strong>Questions:</strong>
1. How related is the age and number of death counts due to COVID-19?
2. Do certain continents have more deaths per million people?
3. Does GDP effect the number of deaths per million people?
4. Is the total number of vaccinations lower on certain continents?

### Part 2 - Data Inspection

Write some code to summarize the datasets.  Think about the following questions:
- What type of data is each variable? (think like a data scientist here, not a computer scientist)
- What is the total size of the data sets?
- What time boundaries are there in the dataset?  IOW, what time frame do they span?
- Are there any missing values in any of the variables? 

Do this with Intentionality.  Don't skimp. 

<strong>Data Types:</strong>
- Categorical: iso_code, continent, location, date
- Numerical: The rest

<strong>Total Dataset Size:</strong>
- Number of rows: 153172
- Number of columns/variables: 67

<strong>Time Frame:</strong>
- Data from 1/15/2020 to 1/8/2022

In [30]:
data_orig.dropna(axis=1)[:5]

iso_code     location        date
0      AFG  Afghanistan  2020-02-24
1      AFG  Afghanistan  2020-02-25
2      AFG  Afghanistan  2020-02-26
3      AFG  Afghanistan  2020-02-27
4      AFG  Afghanistan  2020-02-28

<strong>Missing Values</strong>
- The following variables do <strong>NOT</strong> have missing or null values:
  - iso_code
  - location
  - date
- The rest of the variables have at least on null value

### Part 3 - Data Description

- Create a data description (data dictionary) for your data sets.
    - Describe each variable (4-6 word explanation)
    - If categorical, what levels are present? If the levels are encoded, what do the codes mean?
    - If numeric, provide min, max, median and any other univariate stats you'd like to add in. 
- Where appropriate, provide histograms or other visualizations to characterize each variable. 

In [51]:
pandas.options.display.max_rows = 65
data_orig.min()

iso_code                                              ABW
location                                      Afghanistan
date                                           2020-01-01
total_cases                                           1.0
new_cases                                        -74347.0
new_cases_smoothed                                -6223.0
total_deaths                                          1.0
new_deaths                                        -1918.0
new_deaths_smoothed                              -232.143
total_cases_per_million                             0.001
new_cases_per_million                           -3125.829
new_cases_smoothed_per_million                   -272.971
total_deaths_per_million                              0.0
new_deaths_per_million                            -75.911
new_deaths_smoothed_per_million                   -10.844
reproduction_rate                                   -0.03
icu_patients                                          0.0
icu_patients_p

In [52]:
pandas.options.display.max_rows = 65
data_orig.max()

iso_code                                               ZWE
location                                          Zimbabwe
date                                            2022-01-08
total_cases                                    305191603.0
new_cases                                        2879121.0
new_cases_smoothed                             2265657.571
total_deaths                                     5484782.0
new_deaths                                         18062.0
new_deaths_smoothed                              14704.714
total_cases_per_million                         341391.525
new_cases_per_million                            51427.491
new_cases_smoothed_per_million                     7842.85
total_deaths_per_million                          6085.808
new_deaths_per_million                             453.772
new_deaths_smoothed_per_million                    144.167
reproduction_rate                                     5.92
icu_patients                                       28891

In [53]:
pandas.options.display.max_rows = 5

<strong>Categorical</strong>

| Column Name | Values | Column Description|
| --- | --- | --- |
| iso_code | All ISO codes for each country (text) | The ISO code for a country |
| continent | All continent  names (text) | The continent the country is on |
| location | All country names (text) | The name of the country |
| date | Dates from 1/15/2020 to 1/8/2022 (text) | Date the data was collected |
| tests_units | tests performed / people tested / units unclear / samples tested | Number of data reporting units |

<strong>Numerical</strong>

| Column Name | Minimum Value | Maximum Value |Column Description|
| --- | --- | --- | --- |
| total_cases | 1 | 305191603 | Total COVID-19 cases |
| new_cases | -74347 | 2879121 | Total new COVID-19 cases |
| new_cases_smoothed | -6223 | 2265657.571 | new_cases smoothed to 7 days |
| total_deaths | 1 | 5484782 | Total deaths due to COVID-19 |
| new_deaths | -1918 | 18062 |Total new deaths due to COVID-19 |
| new_deaths_smoothed | -232.143 | 14704.714 | new_deaths column smoothed to 7 days |
| total_cases_per_million | 0.001 | 341391.525 | Total COVID_19 deaths per million people |
| new_cases_per_million | -3125.829 | 51427.491 | Total new cases per million people |
| new_cases_smoothed_per_million | -272.971 | 7842.85 | new_cases_per_million column smoothed to 7 days |
| total_deaths_per_million | 0 | 6085.808 | Total deaths per million people |
| new_deaths_per_million | -75.911 | 453.772 | Total new deaths per million people |
| new_deaths_smoothed_per_million | -10.844 | 144.167 | new_deaths_per_million smoothed to 7 days |
| reproduction_rate | -0.03 | 5.92 | Estimated reproduction rate of COVID-19 |
| icu_patients | 0 | 28891.0 | Number of ICU patients on collection day |
| icu_patients_per_million | 0 | 176.791 | Number of ICU patients per million people |
| hosp_patients | 0 | 133268.0 | Number of hospitalized patients |
| hosp_patients_per_million | 0 | 1544.082 | Number of hospitalized patients per million people |
| weekly_icu_admissions | 0 | 4838.0 | Number of ICU patients on collection week |
| weekly_icu_admissions_per_million | 0 | 221.212 | Number of weekly ICU patients per million people |
| weekly_hosp_admissions | 0 | 138761 | Number of hospitalized people on collection week |
| weekly_hosp_admissions_per_million | 0 | 645.808 | Number of weekly hospitalized pateints per million people |
| new_tests | 1 | 3740296 | New tests for COVID-19 |
| total_tests | 0 | 724212456 | Total tests for COVID-19 |
| total_tests_per_thousand | 0 | 22351.573 | Total tests for COVID-19 per 1,000 people |
| new_tests_per_thousand | 0 | 534.013 | New tests for COVID-19 per 1,000 people |
| new_tests_smoothed | 0 | 6274804 | new_tests smoothed to 7 days |
| new_tests_smoothed_per_thousand | 0 | 132.639 | New COVID-19 tests smoothed per 1,000 people |
| positive_rate | 0 | 0.9704 | 7-day average of positive COVID-19 test cases per total tests |
| tests_per_case | 1 | 50000 | 7-day average of total COVID-19 tests per positive case |
| total_vaccinations | 0 | 9421128533 | Total COVID-19 vaccines administered |
| people_vaccinated | 0 | 4655633325 | Total people who received >= 1 vaccine |
| people_fully_vaccinated | 1 | 3934989809 | Total people who are fully vaccinated |
| total_boosters | 1 | 776881247 | Total number of boosters administered |
| new_vaccinations | 0 | 55062967 | Total new vaccinations administered |
| new_vaccinations_smoothed | 0 | 43026538 | new_vaccinations smoothed to 7 days |
| total_vaccinations_per_hundred | 0 | 322.27 | Total vaccinations administered per 100 people |
| people_vaccinated_per_hundred | 0 | 122.28 | Total people vaccinated per 100 people |
| people_fully_vaccinated_per_hundred | 0 | 118.96 | Total fully vaccinated per 100 people |
| total_boosters_per_hundred | 0 | 81.03 | Total boosters given per 100 people |
| new_vaccinations_smoothed_per_million | 0 | 117497 | new vaccinations given per million people smoothed to 7 days |
| new_people_vaccinated_smoothed | 0 | 100594085 | new_people_vaccinated smoothed to 7 days |
| new_people_vaccinated_smoothed_per_hundred | 0 | 11.75 | New people vaccinated per 100 people smoothed to 7 days |
| stringency_index | 0 | 100 | Government Response Stringency Index |
| population | 47 | 7874965730 | Population of the country |
| population_density | 0.137 | 20546.766 | Number of people divided by land area |
| median_age | 15.1 | 48.2 | Median age of the population |
| aged_65_older | 1.144 | 27.049 | Percent of population older than 65 |
| aged_70_older | 0.526 | 18.493 | Percent of population older than 70 |
| gdp_per_capita | 661.24 | 116935.6 | Gross domestic product at purchasing power parity |
| extreme_poverty | 0.1 | 77.6 | Percent of people living in extreme poverty |
| cardiovasc_death_rate | 79.37 | 724.417 | Death rate from cardiovascular disease |
| diabetes_prevalence | 0.99 | 30.53 | Diabetes prevalence percentage |
| female_smokers | 0.1 | 44.0 | Percent of women who smoke |
| male_smokers | 7.7 | 78.1 | Percent of men who smoke |
| handwashing_facilities | 1.188 | 100 | Percent of people with a handwashing facility |
| hospital_beds_per_thousand | 0.1 | 13.8 | Hospital beds per 1,000 people |
| life_expectancy | 53.28 | 86.75 | Life expectancy at birth in 2019 |
| human_development_index | 0.394 | 0.957 | A composite index measuring average achievement |
| excess_mortality_cumulative_absolute | -31959.4 | 1043824.3 | Absolute cumulative excess mortality rate |
| excess_mortality_cumulative | -28.45 | 115.0 | Cumulative excess mortality rate |
| excess_mortality | -95.92 | 374.34 | Percent excess mortality rate |
| excess_mortality_cumulative_per_million | -1745.051271 | 7912.067517 | Cumulative excess mortality rate per million people |

### Part 4 - Merge the data

Now that you have a better feel for each of your two (or three, for the 7394 students) data sets, it is time to merge them. Describe your strategy for merging the data sets and then actually perform the merge.  

Develop a strategy for verifying that the data is properly merged (hoping and finger-crossing are not valid strategies). 

### Part 5 - Explore Bivariate relationships

- Choose a reasoned set of variables to explore further.  You don't have to explore all possible pairs of variables, nor do we want to grade that much.  Choose 7 - 9 variables. One should be a variable that you'd like to predict (target variable) using the others (predictor variables). 
- List your predictor variables
- List your target variable
- Briefly describe why you have chosen these. 

Use any of the available visualizations from Seaborn to explore the relationships between the variables. Explore the relationships among the predictor variables as well as the relationship between each predictor variable and the target variable.  Which of the predictor variables are most strongly related?  Are there any interesting relationships between categorical predictors and numeric predictors?  If there are any dichotomous variables, does that influence any of the relationships? Are the relationships positive or negative? 

Below each plot, you should provide a description and interpretation of the plot.  Make sure to include why the variables in that plot were chosen and what you hope the reader would gain from it as well. 